kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -60,13 +62,14 @@ public class EnvelopeEncryptionFilter<K>
     }
 
     @SuppressWarnings("unchecked")
-    public static <T> CompletionStage<List<T>> join(List<? extends CompletionStage<T>> stages) {
+    public static <T> @CompletesOnThread("T") CompletionStage<List<T>> join(List<? extends @CompletesOnThread("T") CompletionStage<T>> stages) {
```

Comment:
I don't really know how to read this. What is thread `T` how would someone work out what `T` is supposed to represent?

Does annotating a parameter with `@CompletesOnThread` imply that the method needs a stage that would complete on a designated thread? If so shouldn't it actually be forcing that to happen?

Reply:
> I don't really know how to read this. What is thread `T` how would someone work out what `T` is supposed to represent?

There's definitely some hand waving going on, and it's not really helped by my having not defined what the string should contain (beyond "a description"). What we need to say here is that the returned CS will complement on one of the threads which completes the CS in the argument. That's not quite the same thing as "*", which could be any thread at all. And the distinction matters for how the encryption filter where we rely on the fact that all the CS passed in the argument get completed on the filter thread, and therefore we "know" that the returned CS must be completed on the filter thread as well.

> 
> Does annotating a parameter with `@CompletesOnThread` imply that the method needs a stage that would complete on a designated thread? If so shouldn't it actually be forcing that to happen?

That's not really a question about _this PR_, but about our threading model/the rest of our code. 
This PR is only seeking to document. So it's fine to annotate a parameter because it may be useful (to the person reading the method body) to know that detail. I think it's also fine to say "the annotation is part of the contract". That doesn't mean the contract is right/ideal/couldn't be changed, that that's not what this PR is seeking to do.